File: 003 顺序表基本操作实现/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *elems; // 存放的数据
    int length; // 当前数据长度
    int size; // 总长度
} List;

void * createList(int size) {
    if (size < 1) {
        printf("初始化失败\n，size 不能小于1\n");
        return NULL;
    } else {
        List *list = (List *)malloc(sizeof(List));
        list->size = size;
        list->length = 0;
        list->elems = malloc(sizeof(int) * size);
        return list;
    }
}

void insert(List *list, int idx, int value) {
    if (idx < 0 || list->length == list->size || idx > list->length) {
        printf("插入失败\n");
    } else {
        for (int i = list->length - 1; i >= idx; i--) {
            list->elems[i + 1] = list->elems[i];
        }
        list->elems[idx] = value;
        list->length++;
    }
}

void append(List *list, int value) {
    insert(list, list->length, value);
}

void delete(List *list, int idx) {
    if (idx < 0 || idx >= list->length) {
        printf("删除失败\n");
    } else {
        for (int i = idx; i < list->length - 1; i++) {
            list->elems[i] = list->elems[i + 1];
        }
        list->length--;
    }
}

void update(List *list, int idx, int value) {
    if (idx < 0 || idx >= list->length) {
        printf("更新失败\n");
    } else {
        list->elems[idx] = value;
    }
}

int getvalue(List *list, int idx) {
    if (idx < 0 || idx >= list->length) {
        printf("获取失败\n");
        return -1;
    } else {
        return list->elems[idx];
    }
}

int getidx(List *list, int value) {
    for (int i = 0; i < list->length; i++) {
        if (list->elems[i] == value) {
            return i;
        }
    }
    return -1;
}

int length(List *list) {
    return list->length;
}

void clean(List *list) {
    list->length = 0;
}

void destroy(List *list) {
    free(list->elems);
    free(list);
}

void traverse(List *list) {
    for (int i = 0; i < list->length; i++) {
        printf("%d ", list->elems[i]);
    }
    printf("\n");
}


int main() {
    List *list = createList( 10);
    
//    insert(list, 0, 23);
    append(list, 12);
    append(list, 14);
    append(list, 15);
    append(list, 17);
    append(list, 19);
    append(list, 30);
    insert(list, 2, 90);
    append(list, 10);
    append(list, 40);
    append(list, 50);
    
    traverse(list);
    
    delete(list, 0);
    
    traverse(list);
    
    printf("%d \n", getvalue(list, 8));
    
    printf("%d \n", getidx(list, 40));
    
    update(list, 3, 111);
    
    traverse(list);
}

File: 003 顺序表基本操作实现/main.swift
import Foundation

struct List<E: Equatable> {
    var list: [E?]
    var _length: Int
    
    init(size: UInt) {
        list = Array<E?>(repeating: nil, count: Int(size))
        _length = 0
    }
    
    mutating func insert(at idx: UInt, value: E) -> Bool {
        if idx > _length || _length == list.count {
            return false
        } else {
            let range = (Int(idx)..<_length).reversed()
            for i in range {
                list[i + 1] = list[i]
            }
            list[Int(idx)] = value
            _length += 1
            return true
        }
    }
    
    mutating func append(_ value: E) -> Bool {
        return insert(at: UInt(_length), value: value)
    }
    
    mutating func delete(at idx: UInt) -> Bool {
        if idx >= _length {
            return false
        } else {
            let range = Int(idx)..<_length - 1
            for i in range {
                list[i] = list[i + 1]
            }
            _length -= 1
            return true
        }
    }
    
    mutating func update(at idx: UInt, value: E) -> Bool {
        if idx >= _length {
            return false
        } else {
            list[Int(idx)] = value
            return true
        }
    }
    
    func getIndex(of value: E) -> Int? {
        
        for (idx, val) in list[..<_length].enumerated() {
            if value == val {
                return idx
            }
        }
        return nil
    }
    
    func getValue(of idx: UInt) -> E? {
        if idx >= _length {
            return nil
        } else {
            return list[Int(idx)]
        }
    }
    
    func length() -> Int {
        return _length;
    }
    
    mutating func clear() {
        _length = 0
    }
    
    func traverse() {
        for v in list[..<_length] {
            print(v!, terminator: " ")
        }
        print()
    }
}

var list = List<Int>(size: 10)

for _ in 0..<5 {
    list.append(Int.random(in: 0..<20))
}

list.traverse()

list.insert(at: 5, value: 30)
list.traverse()

list.delete(at: 2)
list.traverse()

list.update(at: 1, value: 100)
list.traverse()

print(list.getIndex(of: 100))
print(list.getValue(of: 3))
